### ISSUE TITLE
msgbus as a suitable self-hosted alternative to patchbay

### ISSUE THREAD

**Issue by prologic** [prologic/msgbus#14]
Consider improving magbus to be a suitable alternative to a self-hsoted version of patchbay with similar features.

**Comment by anderspitman** [comment]
I've never used User Agent switching, though I probably overuse query params (something about everything looking like a nail).

**Comment by prologic** [comment]
Can either do User-Agent based behaviour, Content-Type or query parameters :)

**Comment by prologic** [comment]
Had a play with [patchbay](https://patchbay.pub/docs/index.html) some more tonight and wanted to have a go at re-implementing it almost in its entirety with the least amount of code possible.

Here it is:

```go
package main

import (
	"fmt"
	"io/ioutil"
	"net/http"
	"sync"

	log "github.com/sirupsen/logrus"
)

type pubsub struct {
	sync.RWMutex
	subscribers map[string][]chan []byte
}

func newPubSub() *pubsub {
	return &pubsub{
		subscribers: make(map[string][]chan []byte),
	}
}

func (p *pubsub) ServeHTTP(w http.ResponseWriter, r *http.Request) {
	if r.Method == "GET" {
		topic := r.URL.Path

		ch := make(chan []byte, 1)

		p.Lock()
		p.subscribers[topic] = append(p.subscribers[topic], ch)
		p.Unlock()

		if r.URL.Query().Get("persist") == "true" {
			for {
				w.Write(<-ch)
				w.(http.Flusher).Flush()
			}
		} else {
			w.Write(<-ch)
		}
	} else if r.Method == "POST" {
		topic := r.URL.Path

		p.RLock()
		defer p.RUnlock()

		body, err := ioutil.ReadAll(r.Body)
		if err != nil {
			msg := fmt.Sprintf("error reading request body: %s", err)
			log.WithError(err).Error(msg)
			fmt.Fprint(w, msg)
			http.Error(w, "Internal Server Error", http.StatusInternalServerError)
			return
		}

		for _, ch := range p.subscribers[topic] {
			go func(ch chan []byte) {
				ch <- body
			}(ch)
		}
	} else {
		http.Error(w, "Unsupported Method", http.StatusMethodNotAllowed)
	}
}

type queue struct {
	channels map[string]chan []byte
}

func newQueue() *queue {
	return &queue{
		channels: make(map[string]chan []byte),
	}
}

func (q *queue) ServeHTTP(w http.ResponseWriter, r *http.Request) {
	var (
		ch chan []byte
		ok bool
	)

	ch, ok = q.channels[r.URL.Path]
	if !ok {
		ch = make(chan []byte)
		q.channels[r.URL.Path] = ch
	}

	if r.Method == "GET" {
		if r.URL.Query().Get("persist") == "true" {
			for {
				w.Write(<-ch)
				w.(http.Flusher).Flush()
			}
		} else {
			w.Write(<-ch)
		}
	} else if r.Method == "POST" {
		body, err := ioutil.ReadAll(r.Body)
		if err != nil {
			msg := fmt.Sprintf("error reading request body: %s", err)
			log.WithError(err).Error(msg)
			fmt.Fprint(w, msg)
			http.Error(w, "Internal Server Error", http.StatusInternalServerError)
			return
		}

		ch <- body
	} else {
		http.Error(w, "Unsupported Method", http.StatusMethodNotAllowed)
	}
}

func main() {
	http.Handle("/topic/", newQueue())
	http.Handle("/queue/", newPubSub())
	log.Fatal(http.ListenAndServe(":8000", nil))
}
```

Seems to do what patchbay does:

- Multi-Producer-Multi-Consumer
- Pub/Sub
- Seems to support use cases like Poor Man's X, Chat, Notifs, Web Hosting, etc

**Comment by anderspitman** [comment]
@prologic this looks good! You'd need to flesh out some things like handling cancellation, but the major bits are there. Also, I think you have the calls to newQueue and newPubSub swapped. /queue/ is behaving like pubsub. I'm really glad you made this. The production server has shifted more towards focusing on the requester/responder protocol, because it's more general and allows you to tunnel a complete webserver.

If you make a repo for this I'd be happy to link to it. We could even use it to update https://github.com/patchbay-pub/patchbay-simple-server, but I'd want to put together at least a basic set of integration tests to verify compatibility with the prod server. That's on the todo list anyway though.

**Comment by prologic** [comment]
@anderspitman yeah I can see you've made some changes/improvements in the production [patchbay](https://patchbay.pub/) :D To be honest though I'm not really sure I entirely "get" the whole req/res protocol -- I mean I understand its more flexible and you get the full HTTP capabilities; but the original patchbay as you originally presented it looked and felt more like UNIX pipes which is what I liked :)

That being said; I do intend to spin this into a repo; so you're welcome to help steer this!

One thing I'm not so sure on is the "pubsub" pattern whereby the producer is non-blocking. The "patchbay-simple-server" only seems to implement MPMC / Queue though IIIRC and the only other inspired implementation is "duct" which also uses channels.

The problem I find with using buffer channels for "pubsub" is what do you set the buffer to? `1`? `2`? What? Or is it only non-blocking if there are active subscribers?

**Comment by prologic** [comment]
Early repo is here: https://github.com/prologic/conduit

**Comment by anderspitman** [comment]
Yeah I did hesitate for a while deciding whether to implement req/res, since it's not as simple or elegant as the queues. But at the end of the day, it adds a surprisingly small amount of complexity to the implementation, in exchange for a general HTTP server pipe. For example, imagine how you would serve a directory of 1000 photos with queues. You would need 1000 POSTers. With req/res, you can handle that with a single responder (in production you'd probably want to use several workers, but still way less than 1000).

As for pubsub, I'm not sure what you mean. I think your implementation is correct, and similar to the production server. publishers should never block.
